qiskit/extensions/hamiltonian_gate.py

```diff
@@ -0,0 +1,151 @@
+            # a to_matrix method this will call that method.
+            data = data.to_matrix()
+        elif hasattr(data, 'to_operator'):
+            # If input is a BaseOperator subclass this attempts to convert
+            # the object to an Operator so that we can extract the underlying
+            # numpy matrix from `Operator.data`.
+            data = data.to_operator().data
+        # Convert to numpy array in case not already an array
+        data = numpy.array(data, dtype=complex)
+        # Check input is unitary
+        if not is_hermitian_matrix(data):
+            raise ExtensionError("Input matrix is not Hermitian.")
+        # Check input is N-qubit matrix
+        input_dim, output_dim = data.shape
+        num_qubits = int(numpy.log2(input_dim))
+        if input_dim != output_dim or 2**num_qubits != input_dim:
+            raise ExtensionError(
+                "Input matrix is not an N-qubit operator.")
+
+        self._qasm_name = None
+        self._qasm_definition = None
+        self._qasm_def_written = False
+        # Store instruction params
+        super().__init__('hamiltonian', num_qubits, [data, time], label=label)
+
+    def __eq__(self, other):
+        if not isinstance(other, HamiltonianGate):
+            return False
+        if self.label != other.label:
+            return False
+        # Should we match unitaries as equal if they are equal
+        # up to global phase?
+        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=True)
+        times_eq = self.params[1] == other.params[1]
+        return operators_eq and times_eq
+
+    def to_matrix(self):
+        """Return matrix for the unitary."""
+        unbound_t = False
+        if isinstance(self.params[1], ParameterExpression):
+            unbound_t = len(self.params[1].parameters) > 0
+        if isinstance(self.params[1], Number) or not unbound_t:
+            # pylint: disable=no-member
+            return scipy.linalg.expm(1j * self.params[0] * self.params[1])
+        else:
+            raise NotImplementedError("Unable to generate Unitary matrix for "
+                                      "unbound t parameter {}".format(self.params[1]))
+
+    def inverse(self):
+        """Return the adjoint of the unitary."""
+        return self.adjoint()
+
+    def conjugate(self):
+        """Return the conjugate of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), -self.params[1])
+
+    def adjoint(self):
+        """Return the adjoint of the unitary."""
+        return HamiltonianGate(self.params[0], -self.params[1])
+
+    def transpose(self):
+        """Return the transpose of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), self.params[1])
+
+    def _define(self):
+        """Calculate a subcircuit that implements this unitary."""
+        return UnitaryGate(self.to_matrix())
```

Comment:
This shouldn't return anything afaik, but just set `self.definition` instead
```suggestion
    def _define(self):
        """Calculate a subcircuit that implements this unitary."""
        self.definition = UnitaryGate(self.to_matrix()).definition
```

Reply:
Great point, thanks!